core/src/block.rs

```diff
@@ -973,34 +970,41 @@ impl CandidateCommittedBlock {
         HashOf::new(&self.header).transmute()
     }
 
-    /// Return the signatures (as `payload`) that are verified with the `hash` of this block.
+    /// Return signatures that are verified with the `hash` of this block, remove all other.
     #[inline]
-    pub fn verified_signatures(&self) -> impl Iterator<Item = &SignatureOf<Self>> {
-        self.signatures.verified_by_hash(self.hash())
+    pub fn retain_verified_signatures(&mut self) -> impl Iterator<Item = &SignatureOf<Self>> {
+        self.signatures.retain_verified_by_hash(self.hash())
     }
 
     /// Revalidate transaction hashes, verify signatures and produce [`CommittedBlock`]
     ///
     /// # Errors
     /// - If transaction hashes don't match the hashes stored in the block header
     /// - If signatures verification fails
-    pub fn revalidate(self, topology: &Topology) -> Result<CommittedBlock, eyre::Report> {
-        topology
+    pub fn revalidate(
+        mut self,
+        topology: &Topology,
+    ) -> Result<CommittedBlock, (Self, eyre::Report)> {
+        let verified_signatures = self.retain_verified_signatures();
+
+        if topology
             .filter_signatures_by_roles(
                 &[
                     Role::ValidatingPeer,
                     Role::Leader,
                     Role::ProxyTail,
                     Role::ObservingPeer,
                 ],
-                self.verified_signatures(),
+                verified_signatures,
             )
             .len()
-            .ge(&topology.min_votes_for_commit())
-            .then_some(())
-            .ok_or_else(|| {
-                eyre!("The block doesn't have enough valid signatures to be committed.")
-            })?;
+            .lt(&topology.min_votes_for_commit())
+        {
+            return Err((
```

Comment:
Consider `eyre::bail!((self, eyre!("The block doesn't have enough valid signatures to be committed")`. 

I also think that maybe a new error enum would do us a lot of good here. This message is repeated quite a lot, and `eyre::Report` discards any possibility of handling the error downstream.

Reply:
I don't want to make the block part of the error report. A new error variant could make more sense here, but I'd rather leave it as is atm